src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1391,6 +1697,218 @@ public override DataTable GetSchemaTable()
 
         #endregion Schema metadata table
 
+        #region Seeking
+
+        Task SeekToColumn(int column, bool async)
+        {
+            if (_isSequential)
+                return SeekToColumnSequential(column, async);
+            else
+            {
+                SeekToColumnNonSequential(column);
+                return PGUtil.CompletedTask;
+            }
+        }
+
+        void SeekToColumnNonSequential(int column)
+        {
+            if (_column == -1)
+            {
+                // Starting a new row.
+
+                // We currently assume that the row's number of columns is identical to the description's
+#if DEBUG
+                _numColumns = Buffer.ReadInt16();
+                Debug.Assert(RowDescription.NumFields == _numColumns);
+#else
+                Buffer.ReadPosition += 2;
+#endif
```

Comment:
Previously it was in `ProcessDataMessage` and there was no check in the same method. The method was executed once per row without any checks. I think this is a reason of perf degradation.

Reply:
I see what you mean - there's a new `if (_column == -1)` in SeekToColumnNonSequential. That makes sense.

In the sequential case it's possible we need to do I/O to get the number of columns, but in the non-sequential case we don't (because the whole row is buffered). So in the previous design parsing the column count was done in different places (in SeekToColumn for sequential, in ProcessDataMessage for non-sequential) but that's not possible anymore.

Let me think about this and see what I can come up with.

Reply:
Take a look at the latest commit I've pushed, which removes the start-of-row processing from SeekToColumn.